File: public/PublicMedia/static/solutions/go/problem8.go
/*
Purpose: Find the thirteen adjacent digits in the 1000-digit number that have the greatest product.
Note: the number is read in from a txt file
Author: Eben Quenneville
Date: 2023-02-10
*/
package main

import (
	"fmt"
	"os"
)

const NUMDIGITS = 1000
const SEQLENGTH = 13

func check(e error) {
	// error checking for file I/O
	if e != nil {
		panic(e)
	}
}
func readDigits() [NUMDIGITS]int {
	// Read contents of file
	// `data` is an array of bytes
	data, error := os.ReadFile("../8.txt")
	check(error)
	var digits [NUMDIGITS]int
	for i, v := range string(data) {
		// for each 'rune'/char in data
		digits[i] = int(v - '0')
	}
	// Return the digits of the file
	return digits
}

func problem8() int {
	digits := readDigits()
	largest := 0
	for i := 0; i < NUMDIGITS-SEQLENGTH; i++ {
		product := 1
		for j := 0; j < SEQLENGTH; j++ {
			product *= digits[i+j]
		}
		if product > largest {
			largest = product
		}

	}
	return largest
}

func main() {
	solution := problem8()
	if solution != 23514624000 {
		panic("Incorrect solution!")
	}
	fmt.Println("Output:", solution)
}

File: public/PublicMedia/static/solutions/javascript/problem8.js
/*
Purpose: Find the thirteen adjacent digits in the 1000-digit number that have the greatest product.
Note: the number is read in from a txt file 
Author: Eben Quenneville
Date:  2023-02-19
*/
import assert from "node:assert"
import fs from "node:fs"

const SEQLENGTH = 13;
function readDigits(path) {
    try {
        const data = fs.readFileSync(path, 'utf8');
        let digits = data.split('').map(Number);
        return digits
    }
    catch (err) {
        console.log("Error:", err);
    }

}

function problem8() {
    let digits = readDigits("../8.txt");
    let largest = 0;
    for (let i = 0; i < digits.length - SEQLENGTH; i++) {
        let product = 1;
        for (let j = 0; j < SEQLENGTH; j++) {
            product *= digits[i + j];
        }
        if (product > largest) {
            largest = product;
        }
    }
    return largest;
}


const solution = problem8();
assert(solution == 23514624000);
console.log("Output:", solution);
